Firestore/Swift/Source/Codable/FirestoreDecoder.swift

```diff
@@ -0,0 +1,1189 @@
+    
+    private func _superDecoder(forKey key: CodingKey) throws -> Decoder {
+        self.decoder.codingPath.append(key)
+        defer { self.decoder.codingPath.removeLast() }
+        
+        let value: Any = container[key.stringValue] ?? NSNull()
+        return _FirestoreDecoder(referencing: value, at: self.decoder.codingPath)
+    }
+    
+    public func superDecoder() throws -> Decoder {
+        return try _superDecoder(forKey: _FirestoreKey.super)
+    }
+    
+    public func superDecoder(forKey key: Key) throws -> Decoder {
+        return try _superDecoder(forKey: key)
+    }
+}
+
+@available(swift 4.0.0)
+fileprivate struct _FirestoreUnkeyedDecodingContainer : UnkeyedDecodingContainer {
+    // MARK: Properties
+    /// A reference to the decoder we're reading from.
+    private let decoder: _FirestoreDecoder
+    
+    /// A reference to the container we're reading from.
+    private let container: [Any]
+    
+    /// The path of coding keys taken to get to this point in decoding.
+    private(set) public var codingPath: [CodingKey]
+    
+    /// The index of the element we're about to decode.
+    private(set) public var currentIndex: Int
+    
+    // MARK: - Initialization
+    /// Initializes `self` by referencing the given decoder and container.
+    fileprivate init(referencing decoder: _FirestoreDecoder, wrapping container: [Any]) {
+        self.decoder = decoder
+        self.container = container
+        self.codingPath = decoder.codingPath
+        self.currentIndex = 0
+    }
+    
+    // MARK: - UnkeyedDecodingContainer Methods
+    public var count: Int? {
+        return container.count
+    }
+    
+    public var isAtEnd: Bool {
+        return currentIndex >= count!
+    }
+    
+    public mutating func decodeNil() throws -> Bool {
+        guard !isAtEnd else {
+            throw DecodingError.valueNotFound(Any?.self, DecodingError.Context(codingPath: self.decoder.codingPath + [_FirestoreKey(index: self.currentIndex)], debugDescription: "Unkeyed container is at end."))
+        }
+        
+        if container[currentIndex] is NSNull {
+            currentIndex += 1
+            return true
+        } else {
+            return false
+        }
+    }
+    
+    public mutating func decode(_ type: Bool.Type) throws -> Bool {
+        guard !isAtEnd else {
+            throw DecodingError.valueNotFound(type, DecodingError.Context(codingPath: decoder.codingPath + [_FirestoreKey(index: currentIndex)], debugDescription: "Unkeyed container is at end."))
+        }
+        
+        decoder.codingPath.append(_FirestoreKey(index: currentIndex))
+        defer { decoder.codingPath.removeLast() }
+        
+        guard let decoded = try decoder.unbox(container[currentIndex], as: Bool.self) else {
+            throw DecodingError.valueNotFound(type, DecodingError.Context(codingPath: decoder.codingPath + [_FirestoreKey(index: currentIndex)], debugDescription: "Expected \(type) but found null instead."))
```

Comment:
Also `decoder.codingPath` already has `_FirestoreKey(index: currentIndex)` appended to the end of it; this will duplicate the index in the coding path.